### DISCUSSION TITLE
Updating TinyMCE in Orchard CMS (1.9.1)

### DISCUSSION THREAD

**Discussion by tbdeaton** [OrchardCMS/OrchardCore#16476] [Q&A]
Please delete if not allowed - posting here since there's no Discussion section in Orchard CMS.

I'm currently maintaining a project that uses Orchard CMS version 1.9.1. For the rich text editor, we're using the included TinyMCE editor, specifically an older version (4.1.6). Due to security vulnerabilities, I'd like to upgrade TinyMCE to version 7.

My questions are:

- **Compatibility:** Is it possible to upgrade TinyMCE to version 7 while keeping Orchard CMS at version 1.9.1? Are there any known compatibility issues between TinyMCE 7 and Orchard CMS 1.9.1?
- **Custom Plugins:** Our setup includes custom plugins like mediapicker and medialibrary which are integrated with TinyMCE. Will these custom plugins likely need significant modifications to work with TinyMCE 7?
- **Configuration Changes:** What are the key changes in the configuration of TinyMCE from version 4 to version 7 that I should be aware of?
- **Steps and Best Practices:** What are the recommended steps and best practices to follow when upgrading TinyMCE in an existing Orchard CMS project?

**Current Setup:**

- Orchard CMS Version: 1.9.1
- TinyMCE Version: 4.1.6
- Custom Plugins: mediapicker, medialibrary

Any insights, experiences, or guidance on this upgrade path would be greatly appreciated. Thanks!

I've checked GitHub and Stack Overflow, but I can't find any indication that it's possible to update TinyMCE from version 4 with Orchard CMS, much less upgrading all the way to version 7.

**Comment by Piedone** [comment]
@BenedekFarkas perhaps you can help here?

**Comment by BenedekFarkas** [comment]
FYI, the latest O1 (dev branch) code uses TinyMCE version 4.8.3 (running with jQuery 3.3.1), so that's still way beyond the latest version. To answer your specific questions:

- **Compatibility**: TinyMCE is fairly self-contained, no other built-in features (as in "Orchard feature") depend on it specifically, so you should be able to replace the files without breaking the built-in editors. What can cause problems is its dependency on jQuery (at least up to version 4.x), so you might be forced to upgrade that too (which can cause a cascading effect), if TinyMCE 7 is incompatible with the jQuery version shipped with Orchard 1.9.1. From a quick check, it looks to me that TinyMCE 7 doesn't depend on jQuery anymore (but it can be integrated with it) - if that's really the case, then you'll have a much easier time.
- **Custom Plugins**: I can't really speak to that, it depends on TinyMCE 7 entirely, but with the difference being 3 major versions, I'd expect significant breaking changes.
- **Configuration Changes**: That's also up to TinyMCE entirely.
- **Steps and Best Practices**: Since the latest TinyMCE version in O1 is 4.8.3, even if you upgrade your Orchard solution, TinyMCE will be on a lower version, than your desired one, so you'll be fine with modding your TinyMCE module with the updated files. If you choose to upgrade your Orchard solution later, you can just skip the TinyMCE folder. Alternatively, since TinyMCE 7 is so different compared to what ships with Orchard, you can just put the TinyMCE 7 resources into a new, custom module and use that instead of the built-in one.

**Reply by tbdeaton** [reply]
Thank you so much for the thorough answer! I'm going to give it a shot and I'll report back!

**Reply by BenedekFarkas** [reply]
Glad to help, looking forward to hearing how the upgrade goes!
